### PR TITLE
Handle unloaded modules

### LINKED ISSUE #17788: [Watson] Implement and handle module unloaded cases in ExpressionCompiler
This is one of the causes of Watson crash tracked by DevDiv VSO Bug 187805.

For X86, Roslyn is using the metadata pointer after the module instance (and runtime/process) has been unloaded. Since Roslyn’s EC does not listen to module instance unloaded notifications, it is using the pointer after the module instance has been unloaded and closed.
To fix this, we need to make the following changes:
a.	Roslyn: 
i.	ExpressionCompiler (and the vsdconfigxml) should implement IDkmModuleInstanceUnloadNotification, analogous to IDkmModuleModifiedNotification. This ensures that the metadata bytes ptr is alive till the module instance unloaded notification is processed, ensuring that Roslyn will never have a dangling pointer.
ii.	IsBadOrMissingMetadataException should additionally check if the exception is an ObjectDisposedException.
b.	Debugger:
i.	ManagedDM should return E_RPC_DISCONNECTED (translates to ObjectDisposedException) if metadata is requested for a module instance that has been unloaded.

This work item tracks the Roslyn side of this work.

### PR DESCRIPTION
**Pull request by cston**
Fixes #17788.

### CODE REVIEW

**Comment by cston** [comment]
@kevinh-ms, @plnelson, @r-ramesh, @ivanbasov please review.

File: src/ExpressionEvaluator/Core/Source/FunctionResolver/FunctionResolver.cs [hunk 1]
```diff
@@ -192,13 +192,15 @@ private static bool ShouldEnable(DkmProcess process)
             }
         }
 
+        private const uint RPC_E_DISCONNECTED = 0x80010108;
         private const uint COR_E_BADIMAGEFORMAT = 0x8007000b;
         private const uint CORDBG_E_MISSING_METADATA = 0x80131c35;
 
         private static bool IsBadOrMissingMetadataException(Exception e)
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/FunctionResolver/FunctionResolver.cs]: This looks pretty much the same as the method in MetadataUtilities.  Can we just have one?

**Comment by plnelson** [comment]
👍

File: src/ExpressionEvaluator/Core/Source/FunctionResolver/FunctionResolver.cs (hunk 1 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/FunctionResolver/FunctionResolver.cs]: Created shared method.

**Comment by cston** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by cston** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by cston** [comment]
@dotnet-bot retest this please

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                    return true;
+                case CORDBG_E_MISSING_METADATA:
+                    Debug.WriteLine($"Module '{moduleName}' is missing metadata.");
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private static bool IsBadMetadataException(Exception e)
```
Comment (r-ramesh) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs]: Where is IsBadMetadataException used?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        internal const uint RPC_E_DISCONNECTED = 0x80010108;
+        internal const uint COR_E_BADIMAGEFORMAT = 0x8007000b;
+        internal const uint CORDBG_E_MISSING_METADATA = 0x80131c35;
+
+        internal static bool IsBadOrMissingMetadataException(Exception e, string moduleName)
+        {
+            Debug.Assert(moduleName != null);
+            switch (GetHResult(e))
+            {
+                case RPC_E_DISCONNECTED:
```
Comment (r-ramesh) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs]: The Dispatcher intercepts RPC_E_DISCONNECTED and will just throw ObjectDisposedException, whose HRESULT is not RPC_E_DISCONNECTED.
Instead, I think you will need to check `if (e is ObjectDisposedException)`

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs (hunk 2 above)
Comment (r-ramesh) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs]: (could we perhaps exclusively use the above method?)

**Comment by cston** [comment]
@dotnet-bot retest this please

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs (hunk 2 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs]: `IsBadMetadataException` is used in a couple of callers that already have the metadata so we shouldn't hit object disposed or missing metadata exceptions.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs (hunk 3 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities_Exceptions.cs]: Done.

**Pull request merged**